Closes #2321 Updated period vignette (#2357)
pharmaverse/admiral: vignettes/visits_periods.Rmd

always be required. There are two options:

-   Study specific code is used to first derive the variables `PxxSwSDT` and
`PxxSwEDT` in ADSL. Then `create_period_dataset()` and
`derive_vars_joined()` can be used to derive period/subperiod
variables like `ASPER` or `ASPRSDT` in BDS and OCCDS datasets. See an example 
dataset [here](#adsl_example).

-   Study specific code is used to derive a dataset with one
observation per patient, period, and subperiod (see [period
reference dataset](#reference)). Then `derive_vars_period()` can be
used to derive `PxxSwSDT` and `PxxSwEDT` in ADSL and
`derive_vars_joined()` can be used to derive period/subperiod
variables like `ASPER` or `ASPRSDT` in BDS and OCCDS datasets.

It depends on the specific definition of the periods/subperiods which
option works best. If the definition is based on other ADSL variables,
the first option would work best. If the definition is based on
vertically structured data like exposure data (EX dataset), the second
option should be used. This vignette contains examples for both workflows.

## The Period/Phase Reference Dataset {#reference}

> The admiral functions expect separate reference datasets for periods, subperiods,

Reviewer: Did we want to use {} when admiral is written out without the word package after it?  I think this is our convention??

Author: yes agreed, updated.

      VISIT == "Day 1" ~ 1,
      VISIT == "Week 16" ~ 2
    ),
    TRTA = EXTRT,
    APERSDT = convert_dtc_to_dt(EXSTDTC)
  ) %>%
  # Create APEREDT based on start date of next period
  arrange(USUBJID, APERSDT) %>%
  group_by(USUBJID) %>%
  mutate(
    APEREDT = lead(APERSDT) - 1 # one day before start of next period
  ) %>%
  # Tidy up
  ungroup() %>%
  select(-starts_with("EX"), -VISIT)

period_ref
```

The workflow above populates the Period End Date `APEREDT` for all periods 
except the last one. The value of the last period could then be populated 
with the End of Study Date (`EOSDT`) from ADSL, to obtain:

```{r echo=TRUE}
adsl <- tribble(
  ~STUDYID,  ~USUBJID,  ~TRTSDT,           ~TRTEDT,           ~EOSDT,
  "xyz",     "1",       ymd("2022-01-02"), ymd("2022-08-04"), ymd("2022-09-10"),
  "xyz",     "2",       ymd("2023-10-20"), ymd("2024-05-21"), ymd("2024-06-30")
)

period_ref <- period_ref %>%
  left_join(adsl, by = c("STUDYID", "USUBJID")) %>%
  mutate(APEREDT = case_when(
    APERIOD == "1" ~ APEREDT,
    APERIOD == "2" ~ EOSDT
  )) %>%
  select(-EOSDT, -TRTSDT, -TRTEDT)

period_ref
```

### Creating the Period/Phase Reference Dataset from ADSL {#from_adsl}

> If the treatment start and end dates are already included in `ADSL`, we can derive the phase variables directly in `ADSL` and create a phase reference dataset by employing `create_period_dataset()`. Here is an example command to achieve this goal:

Reviewer: We enclose some datasets in backticks and other are not enclosed.  I believe the convention is to enclose datasets and variables in backticks.

  "xyz",     "2",       ymd("2023-10-20"), ymd("2024-05-21"), ymd("2024-06-30")
)

period_ref <- period_ref %>%
  left_join(adsl, by = c("STUDYID", "USUBJID")) %>%
  mutate(APEREDT = case_when(
    APERIOD == "1" ~ APEREDT,
    APERIOD == "2" ~ EOSDT
  )) %>%
  select(-EOSDT, -TRTSDT, -TRTEDT)

period_ref
```

### Creating the Period/Phase Reference Dataset from ADSL {#from_adsl}

If we instead wish to use an ADSL dataset containing treatment start and end dates to create a phase reference dataset, we can do so by employing `create_period_ref()`. Here is an example command to achieve this goal:

```{r, eval = TRUE}
adsl1 <- adsl %>%
  mutate(
    PH1SDT = TRTSDT,
    PH1EDT = TRTEDT + 28,
    APHASE1 = "TREATMENT",
    PH2SDT = TRTEDT + 29,
    PH2EDT = EOSDT,
    APHASE2 = "FUP"
)

phase_ref <- create_period_dataset(
  adsl1,
  new_vars = exprs(PHSDT = PHwSDT, PHEDT = PHwEDT, APHASE = APHASEw)
)

phase_ref
```

## Creating ADSL Period, Subperiod, or Phase Variables {#periods_adsl}

If a period/phase reference dataset is available, the ADSL variables for
periods, subperiods, or phases can be created from this dataset by
calling `derive_vars_period()`.

> For example the phase reference dataset from the previous section can

Author: good suggestion, i switched it around.

If periods, subperiods, or phases are used, the corresponding variables
have to be consistent across all datasets. This can be achieved by
defining the periods, subperiods, or phases once and then using this
definition for all datasets. The definition can be stored in ADSL or in
a separate dataset. In the following examples, this separate dataset is
called period/phase reference dataset depending what data it contains.

## A Note on Study Specific Code

The sections below showcase the available tools in admiral to work with period, 
subperiod and phase variables. However, at some point study specific code will 
always be required. There are two options:

-   Study specific code is used to first derive the variables `PxxSwSDT` and
`PxxSwEDT` in ADSL. Then `create_period_dataset()` and
`derive_vars_joined()` can be used to derive period/subperiod
variables like `ASPER` or `ASPRSDT` in BDS and OCCDS datasets. (See
[example](#adsl_example).)

-   Study specific code is used to derive a dataset with one
observation per patient, period, and subperiod (see [period

> reference dataset](#reference)). Then `derive_vars_period()` can be

Author: Not sure what you mean here @bms63 - sorry. I changed to "See an example dataset here" and linked to the first time the vignette shows one of these types of datasets.

used to derive `PxxSwSDT` and `PxxSwEDT` in ADSL and
`derive_vars_joined()` can be used to derive period/subperiod
variables like `ASPER` or `ASPRSDT` in BDS and OCCDS datasets.

It depends on the specific definition of the periods/subperiods which
option works best. If the definition is based on other ADSL variables,
the first option would work best. If the definition is based on
vertically structured data like exposure data (EX dataset), the second
option should be used. This vignette contains examples for both workflows.

## The Period/Phase Reference Dataset {#reference}

> The period/phase reference dataset should be created according to the design and ADaM needs of the study in question. It should contain one observation per subject and

Reviewer 2: Could this paragraph be moved after the next paragraph. could then just say "See below for an example dataset".

period, subperiod, or phase. See the [next section](#example_period) for an 
example dataset.

> The admiral functions expect separate datasets for periods, subperiods,

Reviewer 2: ```suggestion
The admiral functions expect separate reference datasets for periods, subperiods,
```

and phases. For periods the numeric variable `APERIOD` is expected, for
subperiods the numeric variables `APERIOD` and `ASPER`, and for phases
the numeric variable `APHASEN`.

## Example Creation of the Period/Phase Reference Dataset {#example_period}

Consider a simple crossover study with the following design:

<img src="https://raw.githubusercontent.com/pharmaverse/admiral/2321_period_examples/inst/visit_periods/dummy_study_design.png" width="630" height="252"/>

> Given this design, the ADaM programmer may wish to split the study into two periods:

Reviewer 2: ```suggestion
Given this design, an option could be to split the study into two periods:
```

  "xyz",    "1",       "Week 12", "Drug X", "2022-04-03",
  "xyz",    "1",       "Week 16", "Drug Y", "2022-05-03",
  "xyz",    "1",       "Week 20", "Drug Y", "2022-06-02",
  "xyz",    "1",       "Week 24", "Drug Y", "2022-07-01",
  "xyz",    "1",       "Week 28", "Drug Y", "2022-08-04",
  "xyz",    "2",       "Day 1",   "Drug Y", "2023-10-20",
  "xyz",    "2",       "Week 4",  "Drug Y", "2023-11-21",
  "xyz",    "2",       "Week 8",  "Drug Y", "2023-12-19",
  "xyz",    "2",       "Week 12", "Drug Y", "2024-01-19",
  "xyz",    "2",       "Week 16", "Drug X", "2024-02-20",
  "xyz",    "2",       "Week 20", "Drug X", "2024-03-17",
  "xyz",    "2",       "Week 24", "Drug X", "2024-04-22",
  "xyz",    "2",       "Week 28", "Drug X", "2024-05-21"
)

ex
```

Then to create a period reference dataset, code like this (or similar) would suffice:

```{r echo=TRUE}
period_ref <- ex %>%
  # Select visits marking the start of each period
  filter(VISIT %in% c("Day 1", "Week 16")) %>%
  # Create APERIOD, APERSDT, TRTA based on SDTM counterparts
  mutate(
    APERIOD = case_when(
      VISIT == "Day 1" ~ 1,
      VISIT == "Week 16" ~ 2
    ),

>     APERSDT = ymd(EXSTDTC),

Author: thanks, actioned

    TRTA = EXTRT
  ) %>%
  # Create APEREDT based on start date of next period
  arrange(USUBJID, desc(APERSDT)) %>%
  group_by(USUBJID) %>%
  mutate(

>     APEREDT = lag(APERSDT) - 1 # one day before start of next period

Author: thanks, actioned

  ) %>%
  # Tidy up
  ungroup() %>%
  arrange(USUBJID, APERSDT) %>%
  select(-starts_with("EX"), -VISIT)

period_ref
```

The workflow above populates the Period End Date `APEREDT` for all periods 
except the last one. The value of the last period could then be populated 
with the End of Study Date (`EOSDT`) from ADSL, to obtain:

```{r echo=TRUE}
adsl <- tribble(
  ~STUDYID,  ~USUBJID,  ~EOSDT,
  "xyz",     "1",       ymd("2022-09-10"),
  "xyz",     "2",       ymd("2024-06-30")
)

period_ref <- period_ref %>%
  left_join(adsl, by = c("STUDYID", "USUBJID")) %>%
  mutate(APEREDT = case_when(
    APERIOD == "1" ~ APEREDT,
    APERIOD == "2" ~ EOSDT
  )) %>%
  select(-EOSDT)

period_ref
```

A very similar process could be followed for the phase reference dataset 
for the study phases, this time using Day 1 and Week 28 as reference 

> visits, to obtain:

Author: I'm not sure about this because I like the relative simplicity of the section. If we add yet another workflow variant I think it could confuse things. If anything I would add this in another section? What do you think?

> visits, to obtain:

Author: I gave that a go - please let me know what you think. I also then removed a section that was duplicated. I would be interested in your opinion as to whether the whole vignette reads well now?

## Creating BDS and OCCDS Treatment Variables {#treatment_bds}

If a period reference dataset is available, BDS and OCCDS variables for
If a period/phase reference dataset is available, BDS and OCCDS variables for
treatment can be created by calling `derive_vars_joined()`.

For example the variables `APERIOD` and `TRTA` can be derived from the
period reference dataset defined above.

```{r}
adae <- tribble(
  ~USUBJID, ~ASTDT,
  "1",      "2021-01-05",
  "1",      "2021-02-05",
  "1",      "2021-03-05",
  "1",      "2021-04-05",
  "2",      "2021-02-15",
  "2",      "2021-03-10",
) %>%
  mutate(

>     ASTDT = ymd(ASTDT),

Reviewer 3: I think we should keep this line. It seems to work but it looks strange if `ASTDT` is character instead of date.
